File: minigit-5/ts/src/commands/checkout.ts
// checkout command - Switch branches or restore files

import * as fs from 'fs';
import * as path from 'path';
import { findRepoRoot, normalizePath, ensureDir } from '../utils';
import { readIndex, writeIndex } from '../index-file';
import { getHeadCommit, getCurrentBranch, resolveRef, resolveRevision, setHead, createBranch } from '../refs';
import { walkTree, getTreeFromTreeIsh, getBlob, hashObject, getCommit } from '../objects';
import { IndexEntry } from '../types';

export function checkout(args: string[]): number {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a minigit repository');
    return 1;
  }

  // Parse arguments
  let createBranchFlag = false;
  let startPoint: string | null = null;
  const positionalArgs: string[] = [];
  let afterDashDash = false;
  const pathsAfterDash: string[] = [];

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '--') {
      afterDashDash = true;
    } else if (afterDashDash) {
      pathsAfterDash.push(arg);
    } else if (arg === '-b') {
      createBranchFlag = true;
    } else if (!arg.startsWith('-')) {
      positionalArgs.push(arg);
    }
  }

  // Handle file restore: checkout [<commit>] -- <file>
  if (pathsAfterDash.length > 0) {
    let commitSha: string | null = null;

    // If there's a positional arg before --, it's the commit
    if (positionalArgs.length > 0) {
      try {
        commitSha = resolveRevision(positionalArgs[0], repoRoot);
      } catch (e: any) {
        console.error(`fatal: ${e.message}`);
        return 1;
      }
    }

    return restoreFiles(pathsAfterDash, commitSha, repoRoot);
  }

  // Handle checkout -b <new-branch> [<start-point>]
  if (createBranchFlag) {
    if (positionalArgs.length === 0) {
      console.error('fatal: branch name required');
      return 1;
    }

    const newBranch = positionalArgs[0];
    let startSha: string;

    if (positionalArgs.length > 1) {
      try {
        startSha = resolveRevision(positionalArgs[1], repoRoot);
      } catch (e: any) {
        console.error(`fatal: ${e.message}`);
        return 1;
      }
    } else {
      const headSha = getHeadCommit(repoRoot);
      if (!headSha) {
        console.error('fatal: not a valid object name: HEAD');
        return 1;
      }
      startSha = headSha;
    }

    // Check if branch already exists
    if (resolveRef(`refs/heads/${newBranch}`, repoRoot)) {
      console.error(`fatal: A branch named '${newBranch}' already exists.`);
      return 1;
    }

    // Create branch
    createBranch(newBranch, startSha, repoRoot);

    // Update working tree if switching from different commit
    const currentHeadSha = getHeadCommit(repoRoot);
    if (currentHeadSha !== startSha) {
      const result = updateWorkingTree(startSha, repoRoot);
      if (result !== 0) return result;
    }

    // Update HEAD
    setHead(`ref: refs/heads/${newBranch}`, repoRoot);

    console.log(`Switched to a new branch '${newBranch}'`);
    return 0;
  }

  // Handle checkout <branch|commit>
  if (positionalArgs.length === 0) {
    console.error('fatal: you must specify a branch or commit');
    return 1;
  }

  const target = positionalArgs[0];

  // Check if target is a branch
  const branchSha = resolveRef(`refs/heads/${target}`, repoRoot);
  if (branchSha) {
    // Check for uncommitted changes that would be overwritten
    const conflict = checkForConflictingChanges(branchSha, repoRoot);
    if (conflict) {
      console.error(`error: Your local changes to the following files would be overwritten by checkout:`);
      console.error(`\t${conflict}`);
      console.error('Please commit your changes or stash them before you switch branches.');
      return 1;
    }

    // Update working tree
    const result = updateWorkingTree(branchSha, repoRoot);
    if (result !== 0) return result;

    // Update HEAD
    setHead(`ref: refs/heads/${target}`, repoRoot);

    console.log(`Switched to branch '${target}'`);
    return 0;
  }

  // Try as commit SHA for detached HEAD
  try {
    const sha = resolveRevision(target, repoRoot);

    // Check for conflicts
    const conflict = checkForConflictingChanges(sha, repoRoot);
    if (conflict) {
      console.error(`error: Your local changes to the following files would be overwritten by checkout:`);
      console.error(`\t${conflict}`);
      return 1;
    }

    // Update working tree
    const result = updateWorkingTree(sha, repoRoot);
    if (result !== 0) return result;

    // Detached HEAD
    setHead(sha, repoRoot);

    console.log(`Note: switching to '${target}'.`);
    console.log('');
    console.log("You are in 'detached HEAD' state.");
    console.log(`HEAD is now at ${sha.slice(0, 7)}`);
    return 0;
  } catch (e: any) {
    console.error(`error: pathspec '${target}' did not match any file(s) known to git`);
    return 1;
  }
}

function restoreFiles(paths: string[], commitSha: string | null, repoRoot: string): number {
  const index = readIndex(repoRoot);

  for (const filePath of paths) {
    let content: Buffer;
    let mode: number;

    if (commitSha) {
      // Restore from commit
      try {
        const treeSha = getTreeFromTreeIsh(commitSha, repoRoot);
        const treeFiles = walkTree(treeSha, '', repoRoot);
        const entry = treeFiles.get(filePath);

        if (!entry) {
          console.error(`error: pathspec '${filePath}' did not match any file(s)`);
          return 1;
        }

        content = getBlob(entry.sha, repoRoot);
        mode = parseInt(entry.mode, 8);
      } catch (e: any) {
        console.error(`error: ${e.message}`);
        return 1;
      }
    } else {
      // Restore from index
      const indexEntry = index.entries.find(e => e.path === filePath);
      if (!indexEntry) {
        console.error(`error: pathspec '${filePath}' did not match any file(s)`);
        return 1;
      }

      content = getBlob(indexEntry.sha, repoRoot);
      mode = indexEntry.mode;
    }

    // Write file
    const fullPath = path.join(repoRoot, filePath);
    ensureDir(path.dirname(fullPath));
    fs.writeFileSync(fullPath, content);

    // Set mode
    if ((mode & 0o777) === 0o755) {
      fs.chmodSync(fullPath, 0o755);
    }
  }

  return 0;
}

function checkForConflictingChanges(targetSha: string, repoRoot: string): string | null {
  const index = readIndex(repoRoot);
  const headSha = getHeadCommit(repoRoot);

  // Get files in target tree
  const targetTree = getTreeFromTreeIsh(targetSha, repoRoot);
  const targetFiles = walkTree(targetTree, '', repoRoot);

  // Get files in HEAD tree
  const headFiles = new Map<string, { sha: string; mode: string }>();
  if (headSha) {
    const headTree = getTreeFromTreeIsh(headSha, repoRoot);
    walkTree(headTree, '', repoRoot).forEach((v, k) => headFiles.set(k, v));
  }

  // Check for conflicting changes
  for (const [filePath, targetEntry] of targetFiles) {
    const headEntry = headFiles.get(filePath);
    const fullPath = path.join(repoRoot, filePath);

    // If file differs between HEAD and target
    if (!headEntry || headEntry.sha !== targetEntry.sha) {
      // Check if working tree has local changes
      if (fs.existsSync(fullPath)) {
        const stats = fs.lstatSync(fullPath);
        if (!stats.isDirectory()) {
          const content = stats.isSymbolicLink()
            ? Buffer.from(fs.readlinkSync(fullPath))
            : fs.readFileSync(fullPath);
          const workingSha = hashObject('blob', content);

          // File is modified in working tree
          const indexEntry = index.entries.find(e => e.path === filePath);
          const indexSha = indexEntry?.sha;

          // If working tree differs from both HEAD and index, conflict
          if (headEntry && workingSha !== headEntry.sha && workingSha !== indexSha) {
            return filePath;
          }
        }
      }
    }
  }

  return null;
}

function updateWorkingTree(targetSha: string, repoRoot: string): number {
  const targetTree = getTreeFromTreeIsh(targetSha, repoRoot);
  const targetFiles = walkTree(targetTree, '', repoRoot);

  // Get current HEAD files
  const headSha = getHeadCommit(repoRoot);
  const headFiles = new Map<string, { sha: string; mode: string }>();
  if (headSha) {
    const headTree = getTreeFromTreeIsh(headSha, repoRoot);
    walkTree(headTree, '', repoRoot).forEach((v, k) => headFiles.set(k, v));
  }

  // Remove files that are in HEAD but not in target
  for (const [filePath] of headFiles) {
    if (!targetFiles.has(filePath)) {
      const fullPath = path.join(repoRoot, filePath);
      if (fs.existsSync(fullPath)) {
        fs.unlinkSync(fullPath);
        // Clean up empty directories
        cleanEmptyDirs(path.dirname(fullPath), repoRoot);
      }
    }
  }

  // Create/update files from target
  for (const [filePath, entry] of targetFiles) {
    const fullPath = path.join(repoRoot, filePath);
    ensureDir(path.dirname(fullPath));

    const content = getBlob(entry.sha, repoRoot);

    if (entry.mode === '120000') {
      // Symlink
      if (fs.existsSync(fullPath)) {
        fs.unlinkSync(fullPath);
      }
      fs.symlinkSync(content.toString(), fullPath);
    } else {
      fs.writeFileSync(fullPath, content);
      if (entry.mode === '100755') {
        fs.chmodSync(fullPath, 0o755);
      }
    }
  }

  // Update index to match target
  const index = readIndex(repoRoot);
  index.entries = [];

  for (const [filePath, entry] of targetFiles) {
    const fullPath = path.join(repoRoot, filePath);
    const stats = fs.lstatSync(fullPath);

    const indexEntry: IndexEntry = {
      ctimeSec: Math.floor(stats.ctimeMs / 1000),
      ctimeNsec: Math.floor((stats.ctimeMs % 1000) * 1000000),
      mtimeSec: Math.floor(stats.mtimeMs / 1000),
      mtimeNsec: Math.floor((stats.mtimeMs % 1000) * 1000000),
      dev: stats.dev,
      ino: stats.ino,
      mode: parseInt(entry.mode, 8),
      uid: stats.uid,
      gid: stats.gid,
      size: stats.size,
      sha: entry.sha,
      flags: Math.min(filePath.length, 0xfff),
      path: filePath,
    };

    index.entries.push(indexEntry);
  }

  writeIndex(index, repoRoot);

  return 0;
}

function cleanEmptyDirs(dir: string, repoRoot: string): void {
  if (dir === repoRoot || !dir.startsWith(repoRoot)) return;

  try {
    const entries = fs.readdirSync(dir);
    if (entries.length === 0) {
      fs.rmdirSync(dir);
      cleanEmptyDirs(path.dirname(dir), repoRoot);
    }
  } catch {
    // Directory doesn't exist or not empty
  }
}

File: minigit-5/py/commands/checkout.py
"""checkout command - Switch branches or restore files."""

import sys
import os
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from utils import find_repo_root
from objects import Blob, Tree, Commit
from index import Index, IndexEntry, create_entry_from_file
from refs import (
    resolve_revision, resolve_head, read_ref, write_head, write_ref,
    get_current_branch, get_head_ref
)


def run(args: list[str]) -> int:
    """Switch branches or restore files."""
    repo_root = find_repo_root()
    if repo_root is None:
        print("fatal: not a minigit repository", file=sys.stderr)
        return 1

    # Parse arguments
    create_branch = False
    paths = []
    target = None
    commit_for_files = None

    i = 0
    while i < len(args):
        arg = args[i]
        if arg == '-b':
            create_branch = True
        elif arg == '--':
            paths.extend(args[i + 1:])
            break
        elif target is None:
            target = arg
        elif commit_for_files is None:
            # This could be start-point for -b or commit for -- <paths>
            commit_for_files = arg
        else:
            paths.append(arg)
        i += 1

    if paths:
        # Restore files mode
        if target and commit_for_files is None:
            # checkout <commit> -- <paths>
            # target is the commit, paths are the files
            return restore_files_from_commit(repo_root, target, paths)
        elif target:
            # checkout -- <paths> - restore from index
            # target is actually the first path
            return restore_files_from_index(repo_root, [target] + paths)
        else:
            return restore_files_from_index(repo_root, paths)

    if create_branch:
        if not target:
            print("fatal: branch name required", file=sys.stderr)
            return 1
        return create_and_checkout(repo_root, target, commit_for_files)

    if target:
        return checkout_target(repo_root, target)

    print("fatal: no target specified", file=sys.stderr)
    return 1


def checkout_target(repo_root: Path, target: str) -> int:
    """Checkout a branch or commit."""
    # Check if it's a branch
    branch_sha = read_ref(repo_root, f'refs/heads/{target}')

    if branch_sha:
        # Checkout branch
        return checkout_branch(repo_root, target, branch_sha)
    else:
        # Try as commit SHA
        sha = resolve_revision(repo_root, target)
        if sha:
            return checkout_detached(repo_root, sha)
        else:
            print(f"error: pathspec '{target}' did not match any branch or commit", file=sys.stderr)
            return 1


def checkout_branch(repo_root: Path, branch: str, sha: str) -> int:
    """Switch to a branch."""
    current_sha = resolve_head(repo_root)

    if sha != current_sha:
        # Check for uncommitted changes that would be overwritten
        if has_conflicting_changes(repo_root, sha):
            print("error: your local changes would be overwritten by checkout", file=sys.stderr)
            return 1

        # Update working tree
        update_working_tree(repo_root, sha)

    # Update HEAD to point to branch
    write_head(repo_root, f'ref: refs/heads/{branch}')

    print(f"Switched to branch '{branch}'")
    return 0


def checkout_detached(repo_root: Path, sha: str) -> int:
    """Checkout a specific commit (detached HEAD)."""
    current_sha = resolve_head(repo_root)

    if sha != current_sha:
        if has_conflicting_changes(repo_root, sha):
            print("error: your local changes would be overwritten by checkout", file=sys.stderr)
            return 1

        update_working_tree(repo_root, sha)

    # Set HEAD directly to SHA
    write_head(repo_root, sha)

    print(f"Note: checking out '{sha[:7]}'.")
    print("You are in 'detached HEAD' state.")
    return 0


def create_and_checkout(repo_root: Path, branch: str, start_point: str = None) -> int:
    """Create a new branch and switch to it."""
    from commands.branch import is_valid_branch_name

    if not is_valid_branch_name(branch):
        print(f"fatal: invalid branch name: {branch}", file=sys.stderr)
        return 1

    if read_ref(repo_root, f'refs/heads/{branch}'):
        print(f"fatal: a branch named '{branch}' already exists", file=sys.stderr)
        return 1

    # Get starting point
    if start_point:
        sha = resolve_revision(repo_root, start_point)
        if sha is None:
            print(f"fatal: not a valid object name: '{start_point}'", file=sys.stderr)
            return 1
    else:
        sha = resolve_head(repo_root)
        if sha is None:
            print("fatal: not a valid object name: 'HEAD'", file=sys.stderr)
            return 1

    current_sha = resolve_head(repo_root)
    if sha != current_sha:
        if has_conflicting_changes(repo_root, sha):
            print("error: your local changes would be overwritten by checkout", file=sys.stderr)
            return 1
        update_working_tree(repo_root, sha)

    # Create branch ref
    write_ref(repo_root, f'refs/heads/{branch}', sha)

    # Update HEAD
    write_head(repo_root, f'ref: refs/heads/{branch}')

    print(f"Switched to a new branch '{branch}'")
    return 0


def restore_files_from_index(repo_root: Path, paths: list) -> int:
    """Restore files from index."""
    index = Index.read(repo_root)

    for path in paths:
        path = path.replace(os.sep, '/')
        if path in index.entries:
            entry = index.entries[path]
            restore_file(repo_root, path, entry.sha)
        else:
            print(f"error: pathspec '{path}' did not match any file(s) known to git", file=sys.stderr)

    return 0


def restore_files_from_commit(repo_root: Path, commit_ref: str, paths: list) -> int:
    """Restore files from a specific commit."""
    sha = resolve_revision(repo_root, commit_ref)
    if sha is None:
        print(f"fatal: not a valid object name: '{commit_ref}'", file=sys.stderr)
        return 1

    commit = Commit.read(sha, repo_root)
    tree_files = get_tree_files(repo_root, commit.tree_sha)

    for path in paths:
        path = path.replace(os.sep, '/')
        if path in tree_files:
            restore_file(repo_root, path, tree_files[path][0])

            # Also update index
            index = Index.read(repo_root)
            file_path = repo_root / path
            mode = tree_files[path][1]
            entry = create_entry_from_file(file_path, path, tree_files[path][0], mode)
            index.add_entry(entry)
            index.write(repo_root)
        else:
            print(f"error: pathspec '{path}' did not match any file(s)", file=sys.stderr)

    return 0


def restore_file(repo_root: Path, rel_path: str, sha: str):
    """Restore a single file from blob SHA."""
    blob = Blob.read(sha, repo_root)
    file_path = repo_root / rel_path
    file_path.parent.mkdir(parents=True, exist_ok=True)
    file_path.write_bytes(blob.data)


def has_conflicting_changes(repo_root: Path, target_sha: str) -> bool:
    """Check if there are uncommitted changes that would conflict."""
    index = Index.read(repo_root)
    target_commit = Commit.read(target_sha, repo_root)
    target_files = get_tree_files(repo_root, target_commit.tree_sha)

    # Check each indexed file
    for name, entry in index.entries.items():
        file_path = repo_root / name

        if name in target_files:
            target_blob_sha = target_files[name][0]

            # If file is modified in working tree and target differs
            if file_path.exists():
                if file_path.is_symlink():
                    content = os.readlink(file_path).encode()
                else:
                    content = file_path.read_bytes()

                work_blob = Blob(content)

                if work_blob.sha != entry.sha and target_blob_sha != entry.sha:
                    # Working tree differs from index, and target differs from index
                    return True
        else:
            # File doesn't exist in target - check if modified
            if file_path.exists():
                if file_path.is_symlink():
                    content = os.readlink(file_path).encode()
                else:
                    content = file_path.read_bytes()
                work_blob = Blob(content)
                if work_blob.sha != entry.sha:
                    return True

    return False


def update_working_tree(repo_root: Path, target_sha: str):
    """Update working tree to match target commit."""
    index = Index.read(repo_root)
    current_files = set(index.entries.keys())

    target_commit = Commit.read(target_sha, repo_root)
    target_files = get_tree_files(repo_root, target_commit.tree_sha)

    # Remove files not in target
    for name in current_files:
        if name not in target_files:
            file_path = repo_root / name
            if file_path.exists():
                file_path.unlink()
                # Clean up empty directories
                try:
                    file_path.parent.rmdir()
                except:
                    pass
            index.remove_entry(name)

    # Add/update files from target
    for name, (sha, mode) in target_files.items():
        blob = Blob.read(sha, repo_root)
        file_path = repo_root / name
        file_path.parent.mkdir(parents=True, exist_ok=True)

        if mode == 0o120000:
            # Symlink
            if file_path.exists() or file_path.is_symlink():
                file_path.unlink()
            os.symlink(blob.data.decode(), file_path)
        else:
            file_path.write_bytes(blob.data)
            if mode == 0o100755:
                file_path.chmod(0o755)

        # Update index
        entry = create_entry_from_file(file_path, name, sha, mode)
        index.add_entry(entry)

    index.write(repo_root)


def get_tree_files(repo_root: Path, tree_sha: str, prefix: str = '') -> dict:
    """Recursively get all files from a tree."""
    files = {}
    tree = Tree.read(tree_sha, repo_root)

    for entry in tree.entries:
        full_path = f"{prefix}{entry.name}" if prefix else entry.name
        if entry.mode == 0o40000:
            files.update(get_tree_files(repo_root, entry.sha, full_path + '/'))
        else:
            files[full_path] = (entry.sha, entry.mode)

    return files
